src/oop/user_defined_classes/hard/q3/MainTest.java

```diff
@@ -0,0 +1,78 @@
+package oop.user_defined_classes.hard.q3;
+
+import global.ObjectTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MainTest {
+    // Java
+    private final String testClassName = "Cupboard";
+    private final String intAttributeName1 = "pots";
+    private final String booleanAttributeName1 = "anyClean";
+    private final String methodName = "washComplete";
+    private ObjectTest classInstance;
+
+    @BeforeEach
+    public void setUp() {
+        String testClassString = "oop.user_defined_classes.hard.q3." + testClassName;
+        classInstance = new ObjectTest(testClassString);
+    }
+
+    @Test
+    public void cupboardClassHasCorrectAttributes() {
+        String missingAttributeMessage =
+                "The %s attribute could not be found in your %s class. Please make sure you have added it, it is spelled correctly, and is of the correct type";
+        String incorrectModifierMessage = "Your %s attribute in the %s class does not have the correct visibility modifier.";
+        assertTrue(classInstance.hasField(intAttributeName1, int.class),
+                String.format(missingAttributeMessage, intAttributeName1, testClassName));
+        assertTrue(classInstance.hasModifier(intAttributeName1, "private"),
+                String.format(incorrectModifierMessage, intAttributeName1, testClassName));
+        assertTrue(classInstance.hasField(booleanAttributeName1, boolean.class),
+                String.format(missingAttributeMessage, booleanAttributeName1, testClassName));
+        assertTrue(classInstance.hasModifier(booleanAttributeName1, "private"),
+                String.format(incorrectModifierMessage, booleanAttributeName1, testClassName));
+    }
+
+    @Test
+    public void cupboardClassHasRequiredConstructor() {
+        String missingConstructorMessage = "Your " + testClassName + " class is missing a required constructor.";
+        assertTrue(classInstance.hasConstructor(null), missingConstructorMessage);
```

Comment:
You should also check the constructor modifier here like this
assertTrue(classInstance.hasModifier(classArguments, "public"),
                "Your " + className + " constructor does not have the correct modifier.");